### PR TITLE
Sprite3D Material refactor

### PR DESCRIPTION
**Pull request by super626**
1. Simplify usage of custom shaders. Usage as follow,
   auto mat = Material::createWithFilename("XXX.material");
   sprite3d->setMaterial(mat);
2. Add Sprite3DMaterial, make it much easier to use both built in shader and custom shaders. More built in shaders are needed.

### CODE REVIEW

File: cocos/3d/CCSprite3D.cpp [hunk 1]
```diff
@@ -465,22 +466,21 @@ void Sprite3D::genGLProgramState(bool useLight)
 {
     _shaderUsingLight = useLight;
 
-    std::unordered_map<const MeshVertexData*, GLProgramState*> glProgramestates;
+    std::unordered_map<const MeshVertexData*, Sprite3DMaterial*> materials;
```
Comment (dabingnn) [inline cocos/3d/CCSprite3D.cpp:469]: I would prefer using `genMaterial` function instead of the old one if it does not break too much compatibility.

Comment (super626) [inline cocos/3d/CCSprite3D.cpp:469]: Ok

**Comment by pandamicro** [comment]
@super626 Please fix the conflict.
@dabingnn Please confirm

**Comment by super626** [comment]
[ci rebuild]

**Comment by dabingnn** [comment]
@pandamicro It could be merged.

**Comment by super626** [comment]
@pandamicro Conflict has been resolved

**Comment by pandamicro** [comment]
Please assign an issue for this PR

**Comment by super626** [comment]
fix #13928

**Pull request merged**